### ISSUE TITLE
Emit a diagnostic when we see conflicting declared types between class body and `__init__` (see TODO in `own_instance_member`)

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ruff#15964]

**Comment by smokyabdulrahman** [comment]
I am not sure if this is helpful, but here is what `pyright` reports for the mentioned test case:

```python
class C:
    """
    Diagnostics:
    Declaration "z" is obscured by a declaration of the same name [reportRedeclaration]
    """

    z: int

    def __init__(self) -> None:
        self.x = get_int()
        """
        Diagnostics:
        Declaration "y" is obscured by a declaration of the same name [reportRedeclaration]
        """
        self.y: int = 1

    def other_method(self):
        self.x = get_str()

        self.y: str = "a"

        self.z: str = "a"
```

I am trying to learn/understand the technical stuff behind parsers by reading (books, blogs) and contributing to tools I love that has things to do with parsers (ruff ❤).

is this issue about adding a new rule? or just fixing/refactoring `red-knot` tests mentioned (modify the markdown file and mention the correct error/reveal messages)?

**Comment by sharkdp** [comment]
> is this issue about adding a new rule?

Yes, this is about emitting a new diagnostic (we can probably reuse [conflicting-declarations](https://github.com/astral-sh/ruff/blob/f30fac632685c1edd200005d110cf407992bd69a/crates/red_knot_python_semantic/src/types/diagnostic.rs#L103-L109)) when we see something like this.

We do want to report them in the same places where pyright reports "declaration … is obscured by", but I think we want to use a slightly different wording, as we generally allow re-declarations of symbols. What's not allowed though, is to have multiple visible conflicting declarations. So this is fine:
```py
x: int = 1
x: str = "foo"
```
But this is not:
```py
def _(flag: bool):
    if flag:
        x: str
    else:
        x: int

    x = 1  # error: [conflicting-declarations]
```

Here, we want something similar, but for declarations of attributes.

**Comment by carljm** [comment]
(To be clear, though, this issue is _not_ about adding a new _ruff_ rule, it's about a new red-knot diagnostic. Although they are in the same codebase, red-knot is separate from ruff and shares only the parser/AST with it.)

**Comment by smokyabdulrahman** [comment]
Thank you all for the information. All clear.

I will give this issue a shot.

**Comment by smokyabdulrahman** [comment]
While trying to dive deep into the code, I've found the following comments where I think the `report_lint` should happen.

https://github.com/astral-sh/ruff/blob/cb8b23d60915206d9ad74fd351a36dd0d4ed1231/crates/red_knot_python_semantic/src/types.rs#L4323-L4329

Also, I've found this TODO:
https://github.com/astral-sh/ruff/blob/cb8b23d60915206d9ad74fd351a36dd0d4ed1231/crates/red_knot_python_semantic/src/types.rs#L4224-L4237

However, I believe that `type.rs` doesn't have the context to report the diagnostic for a reason.

Could you please point me to a similar example where a similar rule had been tackled?

My thought process is to implement some logic in:
https://github.com/astral-sh/ruff/blob/cb8b23d60915206d9ad74fd351a36dd0d4ed1231/crates/red_knot_python_semantic/src/types/infer.rs#L899-L904

and:
https://github.com/astral-sh/ruff/blob/cb8b23d60915206d9ad74fd351a36dd0d4ed1231/crates/red_knot_python_semantic/src/types/infer.rs#L929-L934

or to check here if the target is an instance member of class, and check all the symbol declarations in the scope of the class body, by calling `symbol_from_declarations` and report the conflicting declarations if any.
https://github.com/astral-sh/ruff/blob/cb8b23d60915206d9ad74fd351a36dd0d4ed1231/crates/red_knot_python_semantic/src/types/infer.rs#L2121-L2125

**Comment by sharkdp** [comment]
> While trying to dive deep into the code, I've found the following comments where I think the `report_lint` should happen.

This is the place where we would end up if there would be conflicting declarations *in the class body*. This is just one possible source for declarations of an attribute.

> Also, I've found this TODO:

This is the other place, correct. Here, we loop over declarations of attributes in methods. We currently return early, but we probably need to change the loop logic to collect all declared types, and then emit a diagnostic in case there are more than one.

Finally, we also need to consider the option that there is a declaration on the class body, and a conflicting declaration in a method.

--

> However, I believe that `type.rs` doesn't have the context to report the diagnostic for a reason.

We are currently in the process of changing some infrastructure to make this easier. Maybe it would be good to  split this task into two parts: (1) *collect* all conflicting declared types so we would in principle be able to emit a diagnostic — assuming we could do it from anywhere, and (2) actually create a new diagnostic and make the required changes to be able to emit it.

I think you could already start with part 1, and for part 2, it might make sense to wait for those changes to the diagnostics infrastructure.

Let me know if I failed to answer any of the questions that would be required for part 1.

**Comment by smokyabdulrahman** [comment]
Thank you so much, this made it clearer.
I will try and approach part 1.
